File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.java
import java.io.*;
import java.util.*;

public class Main{

    // 필요변수
    static int N;
    static long M;
    static int[] A;

    // 입력
    static void input() throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        // 값 저장
        N = Integer.parseInt(st.nextToken());
        M = Long.parseLong(st.nextToken());

        // 배열 선언
        A = new int[N];

        st = new StringTokenizer(br.readLine());
        // 배열 값 저장
        for (int i = 0; i < N; i++) {
            A[i] = Integer.parseInt(st.nextToken());
        }


        br.close();
    }

    // 풀이
    static void pro() {
        long sum = 0;
        int right = 0, cnt = 0;

        for (int left = 0; left < N; left++) {
            // M 또는 M보다 커질 때까지 더하기
            while (sum < M && right < N) {
                sum += A[right++];
            }

            // M과 같은지 체크
            if (sum == M)   cnt++;
            
            // left를 합에서 빼기
            sum -= A[left];
        }

        System.out.println(cnt);
        
    }
    public static void main(String[] args) throws Exception{
        input();
        pro();
    }
}

File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.cc
#include <iostream>
using namespace std;

int N;
long long M;
int A[10005];

void input() {
    cin >> N >> M;

    for (int i = 0; i < N; i++) {
        cin >> A[i];
    }
}


void pro() {
    int cnt = 0;
    long long sum = 0;
    
    for (int left = 0, right = 0; left < N; left++) {
        // 합이 M보다 작으면 다음 수 더하기
        while (right < N && sum < M) {
            sum += A[right++];
        }

        // M과 같으면 정답에 추가
        if (sum == M)   cnt++;

        // A[left] 제외해서 다음수부터 더하는 거 탐색
        sum -= A[left];
    }

    // 정답 출력
    cout << cnt;
}

int main() {
    // 최적화
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    input();
    pro();
}